engine/src/main/java/org/pentaho/di/shared/SharedObjectsManager.java

```diff
@@ -0,0 +1,223 @@
+  @Override
+  public synchronized void add( String type, SharedObjectInterface sharedObjectInterface )  throws KettleException {
+    populateSharedObjectMap( type );
+    String name = sharedObjectInterface.getName();
+    Node node = toNode( type, sharedObjectInterface );
+    sharedObjectsIO.saveSharedObject( type, name, node );
+
+    // Add it to the map
+    Map<String, SharedObjectInterface> localSharedObjectsMap = sharedObjectsMap.get( type );
+    localSharedObjectsMap.put( name, sharedObjectInterface );
+  }
+
+  /**
+   * Get the list of SharedObjectInterface for the type
+   * @param type The type of sharedObject
+   * @return
+   * @throws KettleException
+   */
+  @Override
+  public List<SharedObjectInterface> getSharedObjects( String type ) throws KettleException {
+    populateSharedObjectMap( type );
+    // Initialize the inner map if it is not initialized for the type
+    Map<String, SharedObjectInterface> localSharedObjectsMap = sharedObjectsMap.computeIfAbsent( type, k -> new HashMap<>() );
+    return localSharedObjectsMap.values().stream().map( sharedObject -> (SharedObjectInterface) sharedObject.makeClone() ).collect( Collectors.toList() );
+
+  }
+
+  /**
+   * Get the SharedObjectInterface object for the type using the name
+   * @param type The type of sharedObject
+   * @param name name of the SharedObject
+   * @return
+   * @throws KettleException
+   */
+  @Override
+  public SharedObjectInterface getSharedObject( String type, String name ) throws KettleException {
+    populateSharedObjectMap( type );
+    Map<String, SharedObjectInterface> localSharedObjectsMap = sharedObjectsMap.get( type );
+    SharedObjectInterface sharedObjectInterface = localSharedObjectsMap.get( name );
+    return sharedObjectInterface == null ? sharedObjectInterface : (SharedObjectInterface) sharedObjectInterface.makeClone();
+  }
+
+  /**
+   * Remove the SharedObjectInterface object for a type
+   * @param type The type of sharedObject
+   * @param sharedObjectInterface SharedObject to remove
+   * @throws KettleException
+   */
+  @Override
+  public synchronized void removeSharedObject( String type, SharedObjectInterface sharedObjectInterface ) throws KettleException {
+    removeSharedObject( type, sharedObjectInterface.getName() );
+  }
+
+
+  @Override
+  public synchronized void removeSharedObject( String type, String sharedObjectName ) throws KettleException {
+    populateSharedObjectMap( type );
+
+    this.sharedObjectsIO.delete( type, sharedObjectName );
+    // Remove the sharedObject from the map
+    Map<String, SharedObjectInterface> localSharedObjectsMap = sharedObjectsMap.get( type );
+    localSharedObjectsMap.remove( sharedObjectName );
+  }
+
+  @Override
+  public synchronized void clear( String type ) throws KettleException {
+    this.sharedObjectsIO.clear( type );
+    // reset the initialized flags so sharedObjectMap can be repopulated
+    dbconn_initialized = false;
+    slaveServer_initialized = false;
+
```

Comment:
should we also clear the sharedObjectsMap? Would release some memory, if nothing else.